### PR TITLE
Treat record positional parameters as properties

### LINKED ISSUE #48310: Suggested positional parameter name should be pascal-cased
**Version Used**:  16.8-p3.2

image

Pascal-cased names should be suggested because these become public properties.

(Also see https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/3168#issuecomment-687880281)

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #48310

### CODE REVIEW

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs [hunk 1]
```diff
@@ -305,7 +308,7 @@ internal static async Task<NameDeclarationInfo> GetDeclarationInfoAsync(Document
                 result = IsLastTokenOfType<LocalFunctionStatementSyntax>(token, semanticModel,
                     typeSyntaxGetter: f => f.ReturnType,
                     modifierGetter: f => f.Modifiers,
-                    possibleDeclarationComputer: GetPossibleLocalDeclarations,
+                    possibleDeclarationComputer: (_, m) => GetPossibleMemberDeclarations(m),
```
Comment (Youssef1313) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs:311]:
```suggestion
                    possibleDeclarationComputer: (_, m) => GetPossibleLocalDeclarations(m),
```

**Review by davidwengier — COMMENTED** [review]
May be a nit, but in my opinion it would be nicer to add an `IsPrimaryConstructorParameter` method to handle this case, and save having to add a parameter to the lambda in this way. It would also make the resulting code self documenting.

**Comment by Youssef1313** [comment]
@davidwengier I think I'd still need the lambda parameter if I will rely on `IsLastTokenOfType`. But I might include a new parameter for `IsLastTokenOfType` that will hold `Func<TSyntaxNode, bool>`. This argument will always be `_ => true` except for the new introduced method `IsPrimaryConstructorParameter`. This is the approach that comes to mind if I introduced the method you suggested. Is there a better approach?

**Comment by davidwengier** [comment]
@Youssef1313 I haven't thought it through completely, so if you think its not possible, or any better, thats cool, I just pictured something like the tuple handling, where you can use `token.GetAncestor<TSyntaxNode>()`, and if the node is a parameter of a primary constructor, then go from there.

**Comment by Youssef1313** [comment]
@davidwengier I made it as a separate method as you suggested and relied on calling `token.GetAncestor`. The only problem with this approach is that GetAncestor is now called twice for the same token. One time in `IsLastTokenOfType` and the other in `IsPrimaryConstructorParameter`. I assume the call is cheap and won't impact performance.

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs [hunk 2]
```diff
@@ -377,6 +381,24 @@ private static bool IsTypeParameterDeclaration(SyntaxToken token, out NameDeclar
                 return false;
             }
 
+            private static bool IsPrimaryConstructorParameter(SyntaxToken token, SemanticModel semanticModel,
+                CancellationToken cancellationToken, out NameDeclarationInfo result)
+            {
+                result = IsLastTokenOfType<ParameterSyntax>(
```
Comment (Youssef1313) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs:387]: By the way, the method name here is very confusing. The name suggests that it returns a bool, but it doesn't.

I would suggest an analyzer to warn for things like this.
Would this be an IDExxxx rule or a CAxxxx rule?

File: src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs [hunk 3]
```diff
@@ -377,6 +381,20 @@ private static bool IsTypeParameterDeclaration(SyntaxToken token, out NameDeclar
...
+                    token, semanticModel,
+                    p => p.Type,
+                    _ => default,
+                    _ => s_propertySyntaxKind,
+                    cancellationToken);
+
+                return result.Type != null &&
+                    token.GetAncestor<ParameterSyntax>().Parent.IsParentKind(SyntaxKind.RecordDeclaration);
```
Comment (davidwengier) [inline src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.DeclarationInfo.cs:395]: This check should come first, and `result` should be default if it fails.

**Comment by Youssef1313** [comment]
Closing/re-opening for fresh build.

**Comment by Youssef1313** [comment]
@davidwengier for reviewing after applying review feedback

**Review by davidwengier — APPROVED** [review]
Thanks!

**Pull request merged**